File: iso27001-phoenix/test/unit/quality_score_test.exs
defmodule Iso27001Phoenix.QualityScoreTest do
  use ExUnit.Case, async: true

  alias Iso27001Phoenix.Infrastructure.Telemetry.QualityScore

  test "perfect inputs yield composite score of 1.0" do
    score = QualityScore.calculate(
      auth_checks_passed:        100,
      auth_checks_total:         100,
      audit_events_recorded:     50,
      audit_events_expected:     50,
      availability:              1.0,
      logs_with_correlation_id:  200,
      total_logs:                200,
      p95_latency_ms:            0.0
    )

    assert Float.round(QualityScore.composite(score), 2) == 1.0
    assert QualityScore.passes_gate?(score)
  end

  test "zero security score fails production gate" do
    score = QualityScore.calculate(
      auth_checks_passed: 0,
      auth_checks_total:  100,
      availability: 1.0
    )

    refute QualityScore.passes_gate?(score)
  end

  test "slo_alert detects p95 latency breach" do
    alert = QualityScore.slo_alert(p95_latency_ms: 300.0, total_requests: 100)
    assert alert.p95_latency_breached
    assert alert.any_breach
  end

  test "slo_alert does not breach within thresholds" do
    alert = QualityScore.slo_alert(p95_latency_ms: 100.0, total_requests: 100)
    refute alert.p95_latency_breached
    refute alert.error_rate_breached
  end
end

File: iso27001-gin/tests/unit/quality_score_test.go
package unit_test

import (
	"testing"

	"github.com/iso27001/gin-blueprint/internal/infrastructure/telemetry"
)

func TestQualityScore_PerfectScore(t *testing.T) {
	score := telemetry.Calculate(telemetry.QualityScoreInput{
		AuthChecksPassed:     100,
		AuthChecksTotal:      100,
		AuditEventsRecorded:  100,
		AuditEventsExpected:  100,
		Availability:         1.0,
		LogsWithCorrelID:     100,
		TotalLogs:            100,
		ObservedP95LatencyMS: 0,
		ObservedP99LatencyMS: 0,
	})

	composite := score.Composite()
	if composite < 0.99 {
		t.Errorf("perfect inputs should yield composite ≈ 1.0, got %f", composite)
	}
	if !score.PassesGate() {
		t.Error("perfect score should pass the production gate")
	}
}

func TestQualityScore_BelowGate(t *testing.T) {
	score := telemetry.Calculate(telemetry.QualityScoreInput{
		AuthChecksPassed:    0,
		AuthChecksTotal:     100,
		Availability:        0.5,
		ObservedP95LatencyMS: 1000,
	})

	if score.PassesGate() {
		t.Errorf("poor score (composite=%f) should fail the production gate", score.Composite())
	}
}

func TestSLOAlert_BreachDetection(t *testing.T) {
	// 5xx rate of 1% — exceeds 0.1% threshold
	alert := telemetry.EvaluateSLO(10, 0, 1000, 0, 0)
	if !alert.ErrorRateBreached {
		t.Error("expected error_rate_breached = true")
	}
	if !alert.AnyBreach() {
		t.Error("expected any_breach = true")
	}
}

func TestSLOAlert_NoBreachOnCleanSystem(t *testing.T) {
	alert := telemetry.EvaluateSLO(0, 0, 1000, 100, 200)
	if alert.AnyBreach() {
		t.Error("expected no breaches on clean system")
	}
}
